persist-tool/add.c

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2019 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "add.h"
+
+typedef struct __PersistStateEntry
+{
+  gchar *name;
+  gchar *value;
+} PersistStateEntry;
+
+FILE *input_file;
+
+gboolean
+check_directory_exists(gchar *path)
+{
+  return g_file_test(path, G_FILE_TEST_EXISTS | G_FILE_TEST_IS_DIR);
+}
+
+PersistStateEntry *
+persist_state_entry_new(void)
+{
+  PersistStateEntry *self = g_new0(PersistStateEntry, 1);
+  return self;
+}
+
+gboolean
+persist_state_parse_entry(PersistStateEntry *self, gchar *entry)
+{
+  gchar *p = strchr(entry, '=');
+  if (!p)
+    {
+      return FALSE;
+    }
+  self->name = g_strndup(entry, p - entry);
+  self->name = g_strstrip(self->name);
+  self->value = g_strdup((p+1));
+  return TRUE;
+}
+
+#define NAME_LENGTH 40
+#define MAX_LINE_LEN 4000
+#define MAX_ENTRY_VALUE_LENGTH (3*MAX_LINE_LEN)
+
+/* This function parses a JSON style string
+ * containing series of integer values.
+ * For example:
+ * { "value": "00 00 00 00 01 02 03 0A" } */
```

Comment:
I tried to add something, but I ran into trouble.
```
$ cat test.txt 
{ "alma" : "01" }
$ ./persist-tool add -o ../var/ test.txt -p test.persist 
{ "alma" : "01" }
(null)	-->>FAILED (error: Invalid entry syntax)
$ 
```

How should test.txt look like? Is it possible to provide more elaborate error message?

Reply:
The syntax you tried is invalid I guess. Even if your test seems to be valid JSON string this tool has only a very limited parser and you need to keep the format strict.
It shall be something like this:
`alma = { "value" : "01" }`

```
~/src/persist-tool/build/install/bin (persist-tool *)$ ./persist-tool add -o ./ test.txt 
alma	-->>OK
```

Reply:
Oh it works indeed, thanks.